mac/Keyman4MacIM/Keyman4MacIM/KMSettingsRepository.m

```diff
@@ -0,0 +1,206 @@
+ * Serves as an abstraction to StandardUserDefaults which is currently used to store application settings.
+ */
+
+#import "KMSettingsRepository.h"
+#import "KMLogs.h"
+
+NSString *const kActiveKeyboardsKey = @"KMActiveKeyboardsKey";
+NSString *const kSelectedKeyboardKey = @"KMSelectedKeyboardKey";
+NSString *const kPersistedOptionsKey = @"KMPersistedOptionsKey";
+
+NSString *const kObsoletePathComponent = @"/Documents/";
+NSString *const kNewPathComponent = @"/Library/Application Support/keyman.inputmethod.Keyman/";
+
+/**
+ * Store the version number of the data model in the UserDefaults with this key.
+ * The first version, 1, is defined to indicate that we are storing the data/keyboards in the Library
+ * directory instead of in the Documents directory.
+ */
+NSString *const kDataModelVersion = @"KMDataModelVersion";
+NSInteger const kVersionStoreDataInLibraryDirectory = 1;
+
+@implementation KMSettingsRepository
+
++ (KMSettingsRepository *)shared
+{
+  static KMSettingsRepository *shared = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    shared = [[KMSettingsRepository alloc] init];
+  });
+  return shared;
+}
+
+- (void)setDataModelVersionIfNecessary {
+  if (![self dataStoredInLibraryDirectory]) {
+    [[NSUserDefaults standardUserDefaults] setInteger:kVersionStoreDataInLibraryDirectory forKey:kDataModelVersion];
+  }
+}
+
+/**
+ * If the selectedKeyboard has not been set, then the settings have not been saved in the UserDefaults.
+ * If this method is called after applicationDidFinishLaunching, then it will always return true.
+ * If called from awakeFromNib, then it will return false when running for the first time.
+ */
+- (BOOL)settingsExist
+{
+  return [[NSUserDefaults standardUserDefaults] objectForKey:kSelectedKeyboardKey] != nil;
+}
+
+/**
+ * For the first numbered version of the data model, the app stores the keyboards under the /Library directory
+ * For versions before version 1, the keyboards were stored under the /Documents directory.
+ */
+- (BOOL)dataStoredInLibraryDirectory
+{
+  return [[NSUserDefaults standardUserDefaults] integerForKey:kDataModelVersion] == kVersionStoreDataInLibraryDirectory;
+}
+
+/**
+ * Determines whether the keyboard data needs to be moved from the old location in ~/Documents to the new location ~/Library...
+ * This is true if
+ * 1) the UserDefaults exist (indicating that this is not a new installation of Keyman) and
+ * 2) the value  for KMStoreKeyboardsInLibraryKey is not set to true
+ */
+- (BOOL)dataMigrationNeeded {
+  BOOL keymanSettingsExist = [self settingsExist];
+  os_log([KMLogs dataLog], "  keyman settings exist: %{public}@", keymanSettingsExist ? @"YES" : @"NO" );
+  
+  BOOL dataInLibrary = [self dataStoredInLibraryDirectory];
+  os_log([KMLogs dataLog], "  data stored in Library: %{public}@", dataInLibrary ? @"YES" : @"NO" );
+  
+  return !(keymanSettingsExist && dataInLibrary);
```

Comment:
Actually, this works, but it just looks wrong because of a poor method name. I'll rename it.

That method is only checking the settings, not attempting to access the actual data. Everything in the KMSettingsRepository class only accesses UserDefaults and not the file system.